### ISSUE TITLE
Auditor fails to submit samples: java.io.IOException: java.security.ProviderException: Failed to generate key pair

### ISSUE THREAD

**Issue by aruhier** [GrapheneOS/os-issue-tracker#1657]
Hi,

**Device**: Pixel 7
**GrapheneOS build**: TD1A.221105.001.2022110800

Since today, the Auditor cannot submit any sample to attestation.app:

```
keystore2: keystore2::error: In generate_key.
keystore2:
keystore2: Caused by:
keystore2:     0: While generating Key with remote provisioned attestation key.
keystore2:     1: In KeystoreSecurityLevel::upgrade_keyblob_if_required_with.
keystore2:     2: In utils::upgrade_keyblob_if_required_with: Calling km_op.
keystore2:     3: Error::Km(ErrorCode(-38))
SubmitSampleJob: submit failure
SubmitSampleJob: java.io.IOException: java.security.ProviderException: Failed to generate key pair.
SubmitSampleJob:       at z1.i.h(Unknown Source:18)
SubmitSampleJob:       at app.attestation.auditor.SubmitSampleJob.a(Unknown Source:119)
SubmitSampleJob:       at z1.l.run(Unknown Source:14)
SubmitSampleJob:       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:463)
SubmitSampleJob:       at java.util.concurrent.FutureTask.run(FutureTask.java:264)
SubmitSampleJob:       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
SubmitSampleJob:       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
SubmitSampleJob:       at java.lang.Thread.run(Thread.java:1012)
SubmitSampleJob: Caused by: java.security.ProviderException: Failed to generate key pair.
SubmitSampleJob:       at android.security.keystore2.AndroidKeyStoreKeyPairGeneratorSpi.generateKeyPairHelper(AndroidKeyStoreKeyPairGeneratorSpi.java:717)
SubmitSampleJob:       at android.security.keystore2.AndroidKeyStoreKeyPairGeneratorSpi.generateKeyPair(AndroidKeyStoreKeyPairGeneratorSpi.java:627)
SubmitSampleJob:       at java.security.KeyPairGenerator$Delegate.generateKeyPair(KeyPairGenerator.java:746)
SubmitSampleJob:       at z1.i.h(Unknown Source:11)
SubmitSampleJob:       ... 7 more
SubmitSampleJob: Caused by: android.security.KeyStoreException: Invalid argument (internal Keystore code: -38 message: In generate_key.
SubmitSampleJob:
SubmitSampleJob: Caused by:
SubmitSampleJob:     0: While generating Key with remote provisioned attestation key.
SubmitSampleJob:     1: In KeystoreSecurityLevel::upgrade_keyblob_if_required_with.
SubmitSampleJob:     2: In utils::upgrade_keyblob_if_required_with: Calling km_op.
SubmitSampleJob:     3: Error::Km(ErrorCode(-38))) (public error code: 10 internal Keystore code: -38)
SubmitSampleJob:       at android.security.KeyStore2.getKeyStoreException(KeyStore2.java:369)
SubmitSampleJob:       at android.security.KeyStoreSecurityLevel.handleExceptions(KeyStoreSecurityLevel.java:57)
SubmitSampleJob:       at android.security.KeyStoreSecurityLevel.generateKey(KeyStoreSecurityLevel.java:145)
SubmitSampleJob:       at android.security.keystore2.AndroidKeyStoreKeyPairGeneratorSpi.generateKeyPairHelper(AndroidKeyStoreKeyPairGeneratorSpi.java:690)
SubmitSampleJob:       ... 10 more
```

## Context

I've configured Auditor to send a sample every 24h to attestation.app. I've reset the data of my phone yesterday (not related to this issue, I wanted to start on a fresh user profile), and have restored a backup from Seedvault. I also installed the update 2022110800 from 2022110600 after the reset.

I give a full context as I don't know if the reset + backup restore is the issue, or if it is the update to 2022110800, or something else.

I've cleaned the data of the Auditor app, tried to reboot, cleaned the old entry from attestation.app.

Thanks

**Comment by thestinger** [comment]
It sounds like you've had GrapheneOS installed and Auditor paired since very early on. We initially used StrongBox on the Pixel 7 as we do on other Pixels before we found an issue after OS security updates where it fails to trigger the updates of the keys. We switched to TEE keys instead of StrongBox keys for new pairings but if you paired before then you'll need to clear the pairing on both ends and redo it to downgrade to TEE for now. That's temporary and the issue will be eventually resolved upstream.

**Comment by thestinger** [comment]
If you're cleared Auditor app data then I don't know why you'd have issues since that would be a new pairing.

Are you perhaps blocking connections to remote provisioning.grapheneos.org with some software that you're using?

**Comment by aruhier** [comment]
Thanks for the quick answer and the great explanation!

`If you're cleared Auditor app data then I don't know why you'd have issues since that would be a new pairing.`

Maybe I missed something: I went into the properties of the Auditor app, forced stop, cleaned the data storage.

`Are you perhaps blocking connections to remote provisioning.grapheneos.org with some software that you're using?`

I don't think so, I have nothing blocking it on my phone, and it can reach remoteprovisioning.grapheneos.org without any issue. I tested it over wifi and 5G. Also, the pairing works, just the sample submission cannot generate a key pair.

**Comment by thestinger** [comment]
Ah, never mind then, this is an expected issue caused by the known bug on the Pixel 7 / Pixel 7 Pro. I misunderstood and thought you were getting this error from remote attestation.

Sample submission does both a TEE sample and StrongBox sample. On the Pixel 7, StrongBox breaks after the first OS security update until factory reset. After an OS update, the keystore requires triggering an update of each key and it's not handled properly for the remote provisioned StrongBox attestation keys. Pixel 7 is the first device to use remote provisioning for StrongBox. This issue will get fixed upstream since we reported it.

Clearing app data won't help because the issue is that your OS has been updated to a new security patch level after th attestation keys in the Titan M2 were provisioned. There's no point in resolving it with a factory reset because it will happen again next month. It won't break Auditor working due to us already working around this by temporarily using the TEE for all new pairings.
